### PR TITLE
cloud support device state

### PR DESCRIPTION
**Pull request by JiaweiGithub**
Signed-off-by: jiawei <jiawei.liu@daocloud.io>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: build/crds/devices/devices_v1beta1_device.yaml [hunk 1]
```diff
@@ -755,6 +755,12 @@ spec:
             description: DeviceStatus reports the device state and the desired/reported
               values of twin attributes.
             properties:
+              last_online:
```
Comment (WillardHu) [inline build/crds/devices/devices_v1beta1_device.yaml:758]: Please use camel case, and add a suffix 'Time' makes the field more explicit, like 'lastOnlineTime'.

File: cloud/pkg/common/messagelayer/util.go [hunk 2]
```diff
@@ -42,6 +42,7 @@ const (
 	ResourceDevice               = "device"
 	ResourceTypeTwinEdgeUpdated  = "twin/edge_updated"
 	ResourceTypeMembershipDetail = "membership/detail"
+	ResourceDeviceStateUpdated   = "/state/update"
```
Comment (WillardHu) [inline cloud/pkg/common/messagelayer/util.go:45]: Not consistent with the logic above:  '/state/update', 'membership/detail'. Should you delete the prefix char '/'?

File: cloud/pkg/devicecontroller/constants/service.go [hunk 3]
```diff
@@ -4,6 +4,7 @@ package constants
 const (
 	ResourceTypeTwinEdgeUpdated  = "twin/edge_updated"
 	ResourceTypeMembershipDetail = "membership/detail"
+	ResourceDeviceStateUpdated   = "/state/update"
```
Comment (WillardHu) [inline cloud/pkg/devicecontroller/constants/service.go:7]: Ditto

File: cloud/pkg/devicecontroller/controller/upstream.go [hunk 4]
```diff
@@ -111,7 +115,67 @@ func (uc *UpstreamController) updateDeviceStatus() {
...
+			deviceStates := &DeviceStatus{Status: cacheDevice.Status}
+			deviceStates.Status.State = msgState.Device.State
+			deviceStates.Status.LastOnline = msgState.Device.LastOnline
+
+			cacheDevice.Status.State = msgState.Device.State
+			cacheDevice.Status.LastOnline = msgState.Device.LastOnline
+			uc.dc.deviceManager.Device.Store(deviceID, cacheDevice)
+			body, err := json.Marshal(deviceStates)
+			if err != nil {
+				klog.Errorf("Failed to marshal device states %v", deviceStates)
```
Comment (WillardHu) [inline cloud/pkg/devicecontroller/controller/upstream.go:149]: The error handling looks the same, but the logs use some error levels and some warn levels. Please make they are consistent.

File: cloud/pkg/devicecontroller/controller/upstream.go [hunk 5]
```diff
@@ -111,7 +115,67 @@ func (uc *UpstreamController) updateDeviceStatus() {
...
+				continue
+			}
+			err = uc.crdClient.DevicesV1beta1().RESTClient().Patch(MergePatchType).Namespace(cacheDevice.Namespace).Resource(ResourceTypeDevices).Name(cacheDevice.Name).Body(body).Do(context.Background()).Error()
```
Comment (WillardHu) [inline cloud/pkg/devicecontroller/controller/upstream.go:152]: Appropriate line breaks are required for this line.

File: cloud/pkg/devicecontroller/controller/upstream.go [hunk 6]
```diff
@@ -111,7 +115,67 @@ func (uc *UpstreamController) updateDeviceStatus() {
...
+				klog.Warningf("Message: %s process failure, get node id failed with error: %s", msg.GetID(), err)
+				continue
+			}
+			resource, err := messagelayer.BuildResourceForDevice(nodeID, "twin", "")
+			if err != nil {
+				klog.Warningf("Message: %s process failure, build message resource failed with error: %s", msg.GetID(), err)
+				continue
+			}
+			resMsg.BuildRouter(modules.DeviceControllerModuleName, constants.GroupTwin, resource, model.ResponseOperation)
+			resMsg.Content = commonconst.MessageSuccessfulContent
```
Comment (WillardHu) [inline cloud/pkg/devicecontroller/controller/upstream.go:171]: Is it expected that all errors interrupt the process but do not return an error result?

File: manifests/charts/cloudcore/crds/devices_v1beta1_device.yaml [hunk 7]
```diff
@@ -755,6 +755,12 @@ spec:
             description: DeviceStatus reports the device state and the desired/reported
               values of twin attributes.
             properties:
+              last_online:
```
Comment (WillardHu) [inline manifests/charts/cloudcore/crds/devices_v1beta1_device.yaml:758]: Ditto

File: pkg/apis/devices/v1beta1/device_instance_types.go [hunk 8]
```diff
@@ -47,6 +47,12 @@ type DeviceStatus struct {
 	// Optional: A passive device won't have twin properties and this list could be empty.
 	// +optional
 	Twins []Twin `json:"twins,omitempty"`
+	// Optional: The state of the device.
+	// +optional
+	State string `json:"state,omitempty"`
+	// Optional: The last time the device was online.
+	// +optional
+	LastOnline string `json:"last_online,omitempty"`
```
Comment (WillardHu) [inline pkg/apis/devices/v1beta1/device_instance_types.go:55]: Ditto

File: cloud/pkg/devicecontroller/controller/upstream.go [hunk 9]
```diff
@@ -111,7 +115,67 @@ func (uc *UpstreamController) updateDeviceStatus() {
...
+			if !ok {
+				klog.Warning("Failed to assert to CacheDevice type")
+				continue
+			}
+			deviceStates := &DeviceStatus{Status: cacheDevice.Status}
+			deviceStates.Status.State = msgState.Device.State
+			deviceStates.Status.LastOnline = msgState.Device.LastOnline
+
+			cacheDevice.Status.State = msgState.Device.State
+			cacheDevice.Status.LastOnline = msgState.Device.LastOnline
```
Comment (wbc6080) [inline cloud/pkg/devicecontroller/controller/upstream.go:145]: I have some confusion here. The assignments of `deviceStates` and `cacheDevice` seem to be the same. Is it possible to use only one variable to describe them? Or to put it another way, can we directly use `cacheDevice.Status=deviceStates.Status` to replace it?

File: cloud/pkg/devicecontroller/types/device.go [hunk 10]
```diff
@@ -106,3 +106,9 @@ type DeviceTwinUpdate struct {
 	BaseMessage
 	Twin map[string]*MsgTwin `json:"twin"`
 }
+
+// DeviceTwinUpdate the struct of device twin update
+type DeviceStateUpdate struct {
```
Comment (wbc6080) [inline cloud/pkg/devicecontroller/types/device.go:111]:
```suggestion
// DeviceStateUpdate the struct of device state update
type DeviceStateUpdate struct {
```

File: pkg/apis/devices/v1beta1/device_instance_types.go [hunk 11]
```diff
@@ -263,9 +269,8 @@ type VisitorConfig struct {
 type Device struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
-
-	Spec   DeviceSpec   `json:"spec,omitempty"`
-	Status DeviceStatus `json:"status,omitempty"`
+	Spec              DeviceSpec   `json:"spec,omitempty"`
+	Status            DeviceStatus `json:"status,omitempty"`
```
Comment (wbc6080) [inline pkg/apis/devices/v1beta1/device_instance_types.go:273]:
```suggestion
	Spec    DeviceSpec   `json:"spec,omitempty"`
	Status  DeviceStatus `json:"status,omitempty"`
```
Why need to insert a lot of spaces?

File: build/crds/devices/devices_v1beta1_device.yaml (hunk 1 above)
Comment (JiaweiGithub) [reply build/crds/devices/devices_v1beta1_device.yaml:758]: This name has been used in many places in the past. If want to change it, may have to change a lot of things. Need to start with the edge sqlite table structure.

image

**Review by fisherxu — COMMENTED** [review]
Plz rebase to the latest master branch :)

File: build/crds/devices/devices_v1beta1_device.yaml (hunk 1 above)
Comment (fisherxu) [reply build/crds/devices/devices_v1beta1_device.yaml:758]: +1 :)

**Review by Shelley-BaoYue — COMMENTED** [review]
/lgtm

**Pull request merged**